core/src/stages/grasp_provider.cpp

```diff
@@ -0,0 +1,206 @@
+ *********************************************************************/
+
+/* Author: Boston Cleek
+   Desc:   Grasp generator stage
+*/
+
+#include <moveit/task_constructor/stages/grasp_provider.h>
+#include <grasping_msgs/GraspPlanningAction.h>
+
+namespace moveit {
+namespace task_constructor {
+namespace stages {
+
+constexpr char LOGNAME[] = "grasp_provider";
+
+GraspProvider::GraspProvider(const std::string& action_name, const std::string& stage_name, double goal_timeout,
+                             double server_timeout)
+  : GeneratePose(stage_name), ActionBase(action_name, false, goal_timeout, server_timeout), found_candidates_(false) {
+	auto& p = properties();
+	p.declare<std::string>("eef", "name of end-effector");
+	p.declare<std::string>("object");
+	p.declare<boost::any>("pregrasp", "pregrasp posture");
+	p.declare<boost::any>("grasp", "grasp posture");
+}
+
+void GraspProvider::composeGoal() {
+	const auto& props = properties();
+	grasping_msgs::GraspPlanningGoal goal;
+	goal.object.name = props.get<std::string>("object");
+	clientPtr_->sendGoal(goal,
+	                     std::bind(&GraspProvider::doneCallback, this, std::placeholders::_1, std::placeholders::_2),
+	                     std::bind(&GraspProvider::activeCallback, this),
+	                     std::bind(&GraspProvider::feedbackCallback, this, std::placeholders::_1));
+
+	ROS_DEBUG_NAMED(LOGNAME, "Goal sent to server to grasp object: %s", goal.object.name.c_str());
+}
+
+bool GraspProvider::monitorGoal() {
+	// monitor timeout
+	const bool monitor_timeout = goal_timeout_ > std::numeric_limits<double>::epsilon();
+	const double timeout_time = ros::Time::now().toSec() + goal_timeout_;
+
+	while (nh_.ok()) {
+		ros::spinOnce();
+
+		// timeout reached
+		if (ros::Time::now().toSec() > timeout_time && monitor_timeout) {
+			clientPtr_->cancelGoal();
+			ROS_ERROR_NAMED(LOGNAME, "Grasp pose generator time out reached");
+			return false;
+		} else if (found_candidates_) {
+			// timeout not reached (or not active) and grasps are found
+			// only way return true
+			break;
+		}
+	}
+
+	return true;
+}
+
+void GraspProvider::activeCallback() {
+	ROS_DEBUG_STREAM_NAMED(LOGNAME, "Generate grasp goal now active");
+	found_candidates_ = false;
+}
+
+void GraspProvider::feedbackCallback(const grasping_msgs::GraspPlanningFeedbackConstPtr& feedback) {
+	found_candidates_ = true;
+
+	// Protect grasp candidate incase feedback is sent asynchronously
+	const std::lock_guard<std::mutex> lock(grasp_mutex_);
+	grasp_candidates_ = feedback->grasps;
+}
+
+void GraspProvider::doneCallback(const actionlib::SimpleClientGoalState& state,
+                                 const grasping_msgs::GraspPlanningResultConstPtr& result) {
+	if (state == actionlib::SimpleClientGoalState::SUCCEEDED) {
+		ROS_DEBUG_STREAM_NAMED(LOGNAME, "Found grasp candidates");
+	} else {
+		ROS_ERROR_NAMED(LOGNAME, "No grasp candidates found (state): %s", clientPtr_->getState().toString().c_str());
+	}
+}
```

Comment:
The reason the mutex is here is to handle multiple feedback messages. After discussing with @rhaschke he was concerned there could be many feedback messages containing grasp samples. The flag is used to indicate that at least one message was recieved.

Reply:
But as soon as we enter the if statement `if (monitorGoal()) {` in ` void GraspProvider::compute() {` we will not update the `grasp_candidates_` since the mutex is locked, and it will get unlocked just before we exit the function which mean next compute will send a new goal, and we only get the chance to handle the first feedback, right?

Reply:
I see that would simply this by removing the mutex. Yeah this does only handle the first feedback message if `compute()`.

Reply:
I would like it to be able to handle multiple feedback messages. Do you see a way that could work?

Reply:
What do you think about storing all the grasps from the feedback and results fields into a container. Then after the done callback flags success the `compute()` function operates over all those grasp candidates?

Reply:
See https://github.com/bostoncleek/moveit_task_constructor/pull/2 for a proposal :grimacing: